README.md
comedi2py
=========

Data acquisition via C, data visualisation via python.


Run these two demo programs by typing:

comedi2py QWTtest

or

comedi2py Simple


debian/changelog
comedi2py (0.5-2precise) precise; urgency=low

  * Added info about the demo python scripts in the README.me file

 -- Bernd Porr <berndporr@f2s.com>  Thu, 26 Apr 2012 20:11:00 +0100


comedi2py (0.5-1precise) precise; urgency=low

  * Initial Release for precise pangolin

 -- Bernd Porr <berndporr@f2s.com>  Thu, 26 Apr 2012 19:17:04 +0100


Broken.py
import sys
from PyQt4 import Qt
import PyQt4.Qwt5 as Qwt
from PyQt4.Qwt5.anynumpy import *


class SimplePlot(Qwt.QwtPlot):

    def __init__(self, *args):
        Qwt.QwtPlot.__init__(self, *args)

        # make a QwtPlot widget
        self.setTitle('ReallySimpleDemo.py')
        self.insertLegend(Qwt.QwtLegend(), Qwt.QwtPlot.RightLegend)
        
        # set axis titles
        self.setAxisTitle(Qwt.QwtPlot.xBottom, 'x -->')
        self.setAxisTitle(Qwt.QwtPlot.yLeft, 'y -->')

        # insert a few curves
        cSin = Qwt.QwtPlotCurve('y = sin(x)')
        cSin.setPen(Qt.QPen(Qt.Qt.red))
        cSin.attach(self)

        cCos = Qwt.QwtPlotCurve('y = cos(x)')
        cCos.setPen(Qt.QPen(Qt.Qt.blue))
        cCos.attach(self)
        
        # make a Numeric array for the horizontal data
        x = arange(0.0, 10.0, 0.1)

        # initialize the data
        cSin.setData(x, sin(x))
        cCos.setData(x, cos(x))

        # insert a horizontal marker at y = 0
        mY = Qwt.QwtPlotMarker()
        mY.setLabel(Qwt.QwtText('y = 0'))
        mY.setLabelAlignment(Qt.Qt.AlignRight | Qt.Qt.AlignTop)
        mY.setLineStyle(Qwt.QwtPlotMarker.HLine)
        mY.setYValue(0.0)
        mY.attach(self)

        # insert a vertical marker at x = 2 pi
        mX = Qwt.QwtPlotMarker()
        mX.setLabel(Qwt.QwtText('x = 2 pi'))
        mX.setLabelAlignment(Qt.Qt.AlignRight | Qt.Qt.AlignTop)
        mX.setLineStyle(Qwt.QwtPlotMarker.VLine)
        mX.setXValue(2*pi)
        mX.attach(self)

        # replot
        self.replot()

    # __init__()

# class Plot


def make():
    demo = SimplePlot()
    demo.resize(500, 300)
    demo.show()
    return demo

# make()


def comedistart(a):
    app = Qt.QApplication(['Bla'])
    demo = make()
    app.exec_()
    global b
    b = 0
    print a

def comedidata(a):
    global b
    b = b + a[0]
    print b

def comedistop():
    print a


QWTtest.py
import sys
from PyQt4 import Qt
import PyQt4.Qwt5 as Qwt
import PyQt4.Qwt5.anynumpy as np

# this is taken from the QWT demos and slightly modified
# to get this scrolling plot



class DAQThermo(Qt.QWidget):

    def __init__(self, *args):
        Qt.QWidget.__init__(self, *args)

        self.thermo = Qwt.QwtThermo(self)
        self.thermo.setOrientation(Qt.Qt.Vertical,Qwt.QwtThermo.LeftScale)
        self.thermo.setFillColor(Qt.Qt.green)

        label = Qt.QLabel("Volt", self)
        label.setAlignment(Qt.Qt.AlignCenter)

        layout = Qt.QVBoxLayout(self)
        layout.setMargin(0)
        layout.addWidget(self.thermo)
        layout.addWidget(label)

        self.setFixedWidth(3*label.sizeHint().width())

    # __init__()

    def setValue(self, value):
        self.thermo.setValue(value)

    # setValue()

    def setRange(self,mi,ma):
        self.thermo.setRange(mi,ma)


class SimplePlot(Qwt.QwtPlot):

    def __init__(self, *args):
        Qwt.QwtPlot.__init__(self, *args)

        global x,y,cSin,samplingrate;

        # set axis titles
        self.setAxisTitle(Qwt.QwtPlot.xBottom, 't/sec -->')
        self.setAxisTitle(Qwt.QwtPlot.yLeft, 'u/Volt -->')

        # insert a few curves
        cSin = Qwt.QwtPlotCurve('y = u/Volt')
        cSin.setPen(Qt.QPen(Qt.Qt.red))
        cSin.attach(self)

        # make a Numeric array for the horizontal data
        x = np.arange(0.0, 500, 1)
        x = x / samplingrate;
        # sneaky way of creating an array of just zeroes
        y = x * 0

        # initialize the data
        cSin.setData(x,y)

        # insert a horizontal marker at y = 0
        mY = Qwt.QwtPlotMarker()
        mY.setLineStyle(Qwt.QwtPlotMarker.HLine)
        mY.setYValue(0.0)
        mY.attach(self)

        # replot
        self.replot()

    # __init__()

    def new_data(self,d):
        global y,x,cSin;
# shift the dat
        ym = y[0:-1]
        y = np.concatenate( ([d], ym ) )
        cSin.setData(x,y)
        self.replot()

# class Plot


def makePlot():
    demo = SimplePlot()
    demo.resize(500, 300)
    demo.show()
    return demo

def makeThermo():
    thermo = DAQThermo()
    thermo.resize(100,400)
    thermo.show()
    return thermo


#########################################################
# functions called by comedi2py

# called once with the samplingrate in Hz
def comedistart(a,minValue,maxValue):
    global demo
    global thermo
    global samplingrate
    samplingrate = a
    demo = makePlot()
    thermo = makeThermo()
    thermo.setRange(minValue,maxValue)

# called every sample
def comedidata(a):
    global demo
    demo.new_data(a[0]);
    thermo.setValue(a[0]);

# called at the end
def comedistop():
    print "That's it!"



comedi2py.h
/**
 * comedi2py.h
 * (c) 2012, Bernd Porr, no warranty, GNU-public license
 * BerndPorr@f2s.com
 * www.berndporr.me.uk
 **/
class Comedi2py;
#ifndef COMEDI2PY_H
#define COMEDI2PY_H

#include <QWidget>
#include <QPainter>
#include <QApplication>
#include <QPushButton>
#include <QCheckBox>
#include <QLayout> 
#include <QTextEdit>
#include <QGroupBox>
#include <QLabel>

#include "comediasync.h"
#include <Python.h>

///////////////////////
/// python

// the 3 python functions are called by comedi2py
#define PY_START_FUNCTION_NAME "comedistart"
#define PY_DATA_FUNCTION_NAME "comedidata"
#define PY_STOP_FUNCTION_NAME "comedistop"

class Comedi2py : public QWidget
{

	Q_OBJECT
		
public:
/**
 * Constructor
 **/
	Comedi2py( QWidget *parent, 
		   int nchannels,
		   int num_of_devices,
		   int requrested_sampling_rate,
		   int py_sampling_rate,
		   const char *fname
		);
	
/**
 * Destructor
 **/
	~Comedi2py();
	
	/**
	 * The widget which contains the graphical plots of the AD-data
	 **/
	ComediAsync* comediAsync;
	
	/**
	 * Text-field: elapsed time
	 **/
	QTextEdit* timeInfoTextEdit;
	
private:
	/**
	 * Button which controls recording
	 **/
	QCheckBox *recPushButton;
	
        /**
	 * Text-field: time between samples
	 **/
	QTextEdit   *tbInfoTextEdit;

private:
        /**
	 * Sampling rate seen by the python callback functions
	 **/
	int pySamplingRate;

        /**
	 * filename of the python file
	 **/
         const char* filename;

///////////////////////////////////////////////
//Python stuff

private:
	 PyObject *pName, *pModule, *pDict;
	 PyObject *pStartFunc,*pStopFunc,*pDataFunc;
	 PyObject *pValue;
	 PyObject *pValueComedidev;
	 PyObject **pList;
	 PyObject *pArgs,*pStartArgs,*pStopArgs;
	 PyObject *pValueSamplingrate;
	 PyObject *pValueMinRange;
	 PyObject *pValueMaxRange;

public:
	 void initPython();
	 void runPython(float **buffer);
	 void closePython();
};


#endif


comediasync.h
/**
 * comediasync.h
 * (c) 2012 Bernd Porr, no warranty, GNU-public license
 * this encapsulates all the comedi async acquisition
 * It samples at the requested sampling rate and then
 * averages by the decimation amount. For example if
 * the sampling rate is 1kHz and the decimation is
 * 100 then the rate python is called is 10Hz.
 * The callback function is "runPython(float** allChannels)"
 * which in turn calls the python function "comedidata(allChannels)".
 **/
class ComediAsync;
#ifndef COMEDIASYNC_H
#define COMEDIASYNC_H

#include <QWidget>
#include <QTextEdit>

#include <comedilib.h>
#include <fcntl.h>

#include "comedi2py.h"

class ComediAsync : public QTextEdit
{
    Q_OBJECT
public:
/**
 * Constructor:
 **/
    ComediAsync( Comedi2py* comedi2pyTmp,
		 int channels = 0,
		 int maxComediDevices = 1,
		 int req_sampling_rate = 1000
	    );
/**
 * Destructor: close the file if necessary
 **/
    ~ComediAsync();


/**
 * Is called by the timer of the window. This causes the drawing and saving
 * of all not yet displayed/saved data.
 **/
    void	timerEvent( QTimerEvent * );

private slots:

private:
    /**
     * file descriptor for /dev/comedi0
     **/
    comedi_t **dev;

 private:
    unsigned int** chanlist;
    
 private:
    int subdevice;

 private:
    comedi_cmd** cmd;

    /**
     * elapsed msec
     **/
    long int nsamples;

private:
    /**
     * pointer to the parent widget which contains all the controls
     **/
    Comedi2py* comedi2py;
  
private:
    /**
     * buffer which adds up the data for averaging
     **/
    float**   adAvgBuffer;

private:
    /**
     * init value for the averaging counter
     **/
    int         tb_init;

private:
    /**
     * counter for the tb. If zero the average is
     * taken from adAvgBuffer and saved into actualAD.
     **/
    int         tb_counter;

private:
    /**
     * the number of channels actually used per comedi device
     **/
    int channels_in_use;

public:
    int getNchannels() {return channels_in_use;};


private:
    /**
     * the max value of the A/D converter
     **/
    lsampl_t* maxdata;

    /**
     * physical range
     **/
    comedi_range** crange;

public:
    float getMinRange(int nDev) {
	    return crange[nDev]->min;
    }

public:
    float getMaxRange(int nDev) {
	    return crange[nDev]->max;
    }

/**
 * Number of detected comedi devices
 **/
    int nComediDevices;

/**
 * The actual sampling rate
 **/
    int sampling_rate;

public:
/**
 * Start the DAQ board(s)
 **/
    void startDAQ();

public:
/**
 * Gets the number of comedi devices actually used here
 **/
    int getNcomediDevices() {return nComediDevices;};

public:
/**
 * Gets the actual sampling rate the boards are running at.
 **/
    int getDAQSamplingRate() {return sampling_rate;};

private:
/**
 * checks for new data and sends it to python
 **/
    void checkForData();

public:
/**
 * sets the decimation factor (averaging over d samples)
 **/
    void setDecimation(int d);

public:
/**
 * gets the sampling rate perceived by the python callback
 **/
    float getCallbackSamplingRate() { 
	    return ((float)sampling_rate)/
		    ((float)tb_init );
    }

};


#endif


Thermocouple.py
import sys
from PyQt4 import Qt
import PyQt4.Qwt5 as Qwt
import PyQt4.Qwt5.anynumpy as np

# Thermocouple application: channel 0 has the thermocouple
# connected to and channel 1 receives the temperture of the cold junction
# check out http://www.linux-usb-daq.co.uk/howto2/thermocouple/

class DAQThermo(Qt.QWidget):

    def __init__(self, *args):
        Qt.QWidget.__init__(self, *args)

        self.thermo = Qwt.QwtThermo(self)
        self.thermo.setOrientation(Qt.Qt.Vertical,Qwt.QwtThermo.LeftScale)
        self.thermo.setFillColor(Qt.Qt.green)

        label = Qt.QLabel("Temperature", self)
        label.setAlignment(Qt.Qt.AlignCenter)

        layout = Qt.QVBoxLayout(self)
        layout.setMargin(0)
        layout.addWidget(self.thermo)
        layout.addWidget(label)

        self.setFixedWidth(3*label.sizeHint().width())

    # __init__()

    def setValue(self, value):
        self.thermo.setValue(value)

    # setValue()

    def setRange(self,mi,ma):
        self.thermo.setRange(mi,ma)


# this is taken from the QWT demos and slightly modified
# to get this scrolling plot
class ScrollingPlot(Qwt.QwtPlot):

    def __init__(self, *args):
        Qwt.QwtPlot.__init__(self, *args)

    def initPlotwindow(self,y,samplingrate):
        self.samplingrate = samplingrate;

        # set axis titles
        self.setAxisTitle(Qwt.QwtPlot.xBottom, 't/sec -->')
        self.setAxisTitle(Qwt.QwtPlot.yLeft, 'temperature/C -->')

        # insert a few curves
        self.cData = Qwt.QwtPlotCurve('y = temperature')
        self.cData.setPen(Qt.QPen(Qt.Qt.red))
        self.cData.attach(self)

        # make a Numeric array for the horizontal data
        self.x = np.arange(0.0, 500, 1)
        self.x = self.x / samplingrate;
        # sneaky way of creating an array of just zeroes
        self.y = self.x * 0 + y

        # initialize the data
        self.cData.setData(self.x,self.y)

        # insert a horizontal marker at y = 0
        mY = Qwt.QwtPlotMarker()
        mY.setLineStyle(Qwt.QwtPlotMarker.HLine)
        mY.setYValue(0.0)
        mY.attach(self)

        # replot
        self.replot()

    # __init__()

    def new_data(self,d):
        # shift the data to create a scrolling dataplotx
        self.y = np.concatenate( ([d], self.y[0:-1] ) )
        self.cData.setData(self.x,self.y)
        self.replot()

# class Plot


# calculate the temperature
def calcTemperature(voltageTheormocouple,temperatureLM35):
# gain of the instrumentation amplifier INA126
    GAIN_INSTR_AMP=((5+80/0.456))
# zero offset of the instrumentation amplifier
    ZERO_INSTR_AMP=(-0.05365)
    return (((voltageTheormocouple-ZERO_INSTR_AMP)/GAIN_INSTR_AMP)/39E-6) + temperatureLM35


def makePlot(samplingrate):
    scrollplot = ScrollingPlot()
    scrollplot.initPlotwindow(0,samplingrate)
    scrollplot.resize(500, 300)
    scrollplot.show()
    return scrollplot

def makeThermo():
    thermo = DAQThermo()
    thermo.resize(100,400)
    thermo.setRange(-20,300)
    thermo.show()
    return thermo


#########################################################
# functions called by comedi2py

# called once with the samplingrate in Hz
def comedistart(samplingrate,minValue,maxValue):
    global scrollplot
    global thermo
    scrollplot = makePlot(samplingrate)
    thermo = makeThermo()

# called every sample
def comedidata(a):
    global scrollplot
    global thermo
    voltage_thermo = a[0]
    temperature_lm35 = a[1] / 10E-3
    temperature = calcTemperature(voltage_thermo,temperature_lm35)
    scrollplot.new_data(temperature);
    thermo.setValue(temperature);

# called at the end
def comedistop():
    print "\n"


comedi2py.cpp
/********************************************************************
 * comedi2py.cpp 
 * License: GNU, GPL
 * (c) 2012, Bernd Porr
 * No Warranty
 ********************************************************************/

#include <sys/ioctl.h>
#include <math.h>

#include <QTimer>
#include <QPainter>
#include <QApplication>
#include <QButtonGroup>
#include <QGroupBox>
#include <QFileDialog>
#include <QSizePolicy>
#include <QTextEdit>
#include <QMainWindow>
#include <QSettings>

#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <qtextedit.h>
#include <qfont.h>


#include <iostream>
#include <fstream>

#include "comediasync.h"
#include "comedi2py.h"

// for the layout
#define MAXROWS 8

Comedi2py::Comedi2py( QWidget *parent, 
		      int nchannels,
		      int num_of_devices,
		      int requrested_sampling_rate,
		      int py_sampling_rate,
		      const char *fname
	)
    : QWidget( parent ) {

	filename = fname;

        comediAsync=new ComediAsync(this,
				    nchannels,
				    num_of_devices,
				    requrested_sampling_rate
		);

	int factual = comediAsync->getDAQSamplingRate();

	int decimation = factual/py_sampling_rate;

	comediAsync->setDecimation(decimation);

	// fonts
	QFont voltageFont("Courier",10);
	QFontMetrics voltageMetrics(voltageFont);

	// this the main layout which contains two sub-windows:
	// the control window and the oscilloscope window
	QVBoxLayout *mainLayout = new QVBoxLayout;
	mainLayout->setSpacing(0);
	setLayout(mainLayout);

	// this is the vertical layout for all the controls
	QHBoxLayout *controlLayout = new QHBoxLayout(0);
	// the corresponding box which contains all the controls
	QGroupBox *controlBox = new QGroupBox ();

	// this is the vertical layout for all the controls
	QHBoxLayout *scopeLayout = new QHBoxLayout(0);
	// the corresponding box which contains all the controls
	QGroupBox *scopeGroup = new QGroupBox ();

	// group for the record stuff
        QGroupBox* recGroupBox = new QGroupBox("python:");
        QHBoxLayout *recLayout = new QHBoxLayout();

        recPushButton = new QCheckBox("&data acquisition");
        recPushButton->setEnabled( true );
        recLayout->addWidget(recPushButton);

        recGroupBox->setLayout(recLayout);
        controlLayout->addWidget(recGroupBox);

       // group for the time base
        QGroupBox *tbgrp = new QGroupBox("Sampling rate");
        QHBoxLayout *tbLayout = new QHBoxLayout;
        QFont tbFont("Courier",12);
        tbFont.setBold(TRUE);
        QFontMetrics tbMetrics(tbFont);

	tbInfoTextEdit = new QTextEdit(tbgrp);
	tbInfoTextEdit->setFont (tbFont);
	QFontMetrics metricsTb(tbFont);
	tbInfoTextEdit->setMaximumSize (3*metricsTb.width("99999 sec")/2,
					3*metricsTb.height()/2);
	tbInfoTextEdit->setReadOnly(TRUE);
	tbInfoTextEdit->setLineWidth(1);
	tbInfoTextEdit->setVerticalScrollBarPolicy( Qt::ScrollBarAlwaysOff );

        QString s;
	float cbsr = comediAsync->getCallbackSamplingRate();
        if (cbsr<1000) {
                s.sprintf( "%5.0f Hz", cbsr);
        } else if (cbsr<1000000) {
                s.sprintf( "%5.0f kHz", cbsr/1000.0);
        } else {
                s.sprintf( "%5.0f Mhz", cbsr/1000000.0);
        }               
        tbInfoTextEdit->setText(s);

	tbLayout->addWidget(tbInfoTextEdit);

	tbgrp->setLayout(tbLayout);
	controlLayout->addWidget(tbgrp);
	controlBox->setLayout(controlLayout);

	comediAsync->setMinimumWidth ( 300 );
	comediAsync->setMinimumHeight ( 200 );

	scopeLayout->addWidget(comediAsync);
	scopeGroup->setLayout(scopeLayout);

	controlBox->setSizePolicy ( QSizePolicy(QSizePolicy::Fixed,
						QSizePolicy::Fixed ) );

	mainLayout->addWidget(controlBox);
	mainLayout->addWidget(scopeGroup);

	initPython();

	comediAsync->startDAQ();
}

Comedi2py::~Comedi2py() {
	delete comediAsync;
	closePython();
}


void Comedi2py::initPython() {
	Py_Initialize();

	// make the interpreter look in the working dir
	PyRun_SimpleString("import sys");
	char *workingdir = get_current_dir_name();
	char tmp[1024];
	sprintf(tmp,"sys.path.append(\"%s\")",workingdir);
	PyRun_SimpleString(tmp);
	free(workingdir);

	// load the module
	pName = PyString_FromString(filename);
	pModule = PyImport_Import(pName);
	Py_DECREF(pName);

	if (pModule == NULL) {
		fprintf(stderr,"Python module not loadable:\n");
		PyErr_Print();
		exit(1);
	}

	pStartFunc = PyObject_GetAttrString(pModule,PY_START_FUNCTION_NAME);
	if (!pStartFunc) {
		fprintf(stderr,
			"function "PY_START_FUNCTION_NAME" not defined\n");
		exit(1);
	}
	if (!PyCallable_Check(pStartFunc)) {
		fprintf(stderr,
			"function "PY_START_FUNCTION_NAME" not callable\n");
	}

	pDataFunc = PyObject_GetAttrString(pModule,PY_DATA_FUNCTION_NAME);
	if (!pDataFunc) {
		fprintf(stderr,
			"function "PY_DATA_FUNCTION_NAME" not defined\n");
		exit(1);
	}
	if (!PyCallable_Check(pDataFunc)) {
		fprintf(stderr,
			"function "PY_DATA_FUNCTION_NAME" not callable\n");
		exit(1);
	}

	pStopFunc = PyObject_GetAttrString(pModule,PY_STOP_FUNCTION_NAME);
	if (!pStopFunc) {
		fprintf(stderr,
			"function "PY_STOP_FUNCTION_NAME" not defined\n");
		exit(1);
	}
	if (!PyCallable_Check(pStopFunc)) {
		fprintf(stderr,
			"function "PY_STOP_FUNCTION_NAME" not callable\n");
		exit(1);
	}

	// sampling rate, min, max
	pStartArgs = PyTuple_New(3);
	double r = (double)(comediAsync-> getCallbackSamplingRate());
	pValueSamplingrate = PyFloat_FromDouble(r);
	double minRange = (double)(comediAsync-> getMinRange(0));
	pValueMinRange = PyFloat_FromDouble(minRange);
	double maxRange = (double)(comediAsync-> getMaxRange(0));
	pValueMaxRange = PyFloat_FromDouble(maxRange);
	PyTuple_SetItem(pStartArgs, 0, pValueSamplingrate);
	PyTuple_SetItem(pStartArgs, 1, pValueMinRange);
	PyTuple_SetItem(pStartArgs, 2, pValueMaxRange);
	PyObject_CallObject(pStartFunc, pStartArgs);
	PyErr_Print();

	int ndev = comediAsync->getNcomediDevices();
	int n = comediAsync->getNchannels();
	pArgs = PyTuple_New(ndev);
	pList = new PyObject*[ndev];
	for (int d = 0; d < ndev; d++) {
		pList[d] = PyList_New(n);
		for (int i = 0; i < n; ++i) {
			pValue = PyFloat_FromDouble((double)0.0);
			if (!pValue) {
				Py_DECREF(pArgs);
				Py_DECREF(pModule);
				fprintf(stderr, "Cannot convert argument\n");
				exit(1);
			}
			PyList_SetItem(pList[d], i, pValue);
		}
		PyTuple_SetItem(pArgs, d, pList[d]);
	}
}

void Comedi2py::runPython(float **buffer) {
	if (recPushButton->checkState()==Qt::Checked) 
	{
		int ndev = comediAsync->getNcomediDevices();
		int n = comediAsync->getNchannels();
		for (int d = 0; d < ndev; d++) {
			pList[d] = PyList_New(n);
			for (int i = 0; i < n; ++i) {
				pValue = PyFloat_FromDouble(
					(double)(buffer[d][i])
					);
				if (!pValue) {
					Py_DECREF(pArgs);
					Py_DECREF(pModule);
					fprintf(stderr,
						"Cannot convert argument\n");
					exit(1);
				}
				PyList_SetItem(pList[d], i, pValue);
			}
			PyTuple_SetItem(pArgs, d, pList[d]);
		}
	pValue = PyObject_CallObject(pDataFunc, pArgs);
	PyErr_Print();
	}
}

void Comedi2py::closePython() {
	pStopArgs = PyTuple_New(0);
	PyObject_CallObject(pStopFunc, pStopArgs);
	if (pStartFunc) Py_XDECREF(pStartFunc);
	if (pDataFunc) Py_XDECREF(pDataFunc);
	if (pStopFunc) Py_XDECREF(pStopFunc);
        if (pModule) Py_DECREF(pModule);
}

// options
///////////////////////////////////////////////////////////////////////////
int main( int argc, char **argv )
{
	int c;
	int num_of_channels = 8;
	int num_of_devices = 1;
	const char *filename = NULL;
	int daq_sampling_rate = 1000;
	int py_sampling_rate = 10;
	const char options_getopt[] = "r:d:c:hs:";
	const char usage[] = "usage:\n"					\
		"%s"							\
		" -c <number of channels>"				\
		" -d <max number of comedi devices>"			\
		" -r <sampling rate of the DAQ card>"			\
		" -s <sampling rate of the python callback>"		\
		" PYTHON_MODULE_NAME (without .py)\n";

	QApplication a( argc, argv );		// create application object

	// we are just interested in the module name to get the settings
	// We skip any other option for now
	while (-1 != (c = getopt(argc, argv,options_getopt)))
	{
		switch (c) {
		case 'c':
		case 'd':
		case 'r':
		case 's':
			break;
		case 'h':
		default:
			printf(usage,argv[0]);
		fprintf(stderr,"Try `man comedi2py' for more information.\n");
		exit(1);
		}
	}

	if (optind < argc) {
		filename = argv[optind];
	} else {
		fprintf(stderr,
		       "No python module specified. "
		       "You need to specify at least your python module.\n");
		printf(usage,argv[0]);
		exit(1);
	}
	// just to be sure we really have a module name
	assert(filename!=NULL);

	QSettings settings(QSettings::IniFormat, 
			   QSettings::UserScope,
			   "USB-DUX",
			   "comedi2py");

	settings.beginGroup(filename);
	num_of_channels = settings.value("num_of_channels",16).toInt();
	num_of_devices = settings.value("num_of_devices",1).toInt();
	daq_sampling_rate = settings.value("daq_sampling_rate",1000).toInt();
	py_sampling_rate = settings.value("py_sampling_rate",10).toInt();
	settings.endGroup();
	
	// rewinding to get the options to override the settings
	optind = 1;
	while (-1 != (c = getopt(argc, argv,options_getopt))) {
		switch (c) {
		case 'c':
			num_of_channels = strtoul(optarg,NULL,0);
			break;
		case 'd':
			num_of_devices = atoi(optarg);
			break;
		case 'r':
			daq_sampling_rate = atoi(optarg);
			break;
		case 's':
			py_sampling_rate = atoi(optarg);
			break;
		}
	}

	Comedi2py comedi2py(0,
			    num_of_channels,
			    num_of_devices,
			    daq_sampling_rate,
			    py_sampling_rate,
			    filename
		);

	comedi2py.show();			// show widget

	int ret = a.exec();			// run event loop
	
	// last window has been closed so we can save the settings
	settings.beginGroup(filename);
	settings.setValue("num_of_channels",num_of_channels);
	settings.setValue("num_of_devices",num_of_devices);
	settings.setValue("daq_sampling_rate",daq_sampling_rate);
	settings.setValue("py_sampling_rate",py_sampling_rate);
	settings.endGroup();

	return ret;
}
